### ISSUE TITLE
Correctly implement timeouts

### ISSUE THREAD

**Issue by transistive** [neo4j-php/neo4j-php-client#102]
Hi @transistive :wave: 

I know this issue is closed, but I'm running into the same problem with the timeout. I'm setting my client as follows
```php
$url = 'bolt://test:test@neo4j:7687?database=test'
$this->client = ClientBuilder::create()
                             ->withDriver('bolt', $url, null, 3.0)
                             ->withDefaultTransactionConfiguration(TransactionConfiguration::default()->withTimeout(3.0))
                             ->withDefaultDriver('bolt')
                             ->build();
```

I know `3.0` is a really low value but I was just testing it out. This works for transactions only
```php
$tsx = $this->client->beginTransaction();
$result = $tsx->run("CALL apoc.util.sleep(10000)");
$tsx->commit();
```
It does timeout after 3 seconds with the error `Connection timeout reached after 3 seconds.` which is expected. However, I can't get it to work with the `run` function, so the following code doesn't timeout
```php
$result = $this->client->run("CALL apoc.util.sleep(10000)");
```
This doesn't timeout for some reason.

I tried debugging the code of the client and it appears that this part of the code is the reason
```php
/**
 * @return UnmanagedTransactionInterface<ResultFormat>
 */
private function beginInstantTransaction(SessionConfiguration $config): TransactionInterface
{
    $connection = $this->acquireConnection(TransactionConfiguration::default(), $config);

    return new BoltUnmanagedTransaction($this->config->getDatabase(), $this->formatter, $connection);
}

/**
 * @throws Exception
 *
 * @return ConnectionInterface<Bolt>
 */
private function acquireConnection(TransactionConfiguration $config, SessionConfiguration $sessionConfig): ConnectionInterface
{
    $timeout = max($this->socketTimeout, $config->getTimeout());

    return $this->pool->acquire($this->uri, $this->auth, $timeout, $this->userAgent, $sessionConfig);
}
```

The `$config->getTimeout()` part always returns `100000` which is the default timeout, so it overrides the passed `3.0` and this happens because it passed to it `TransactionConfiguration::default()` which has a null timeout so it always defaults to `100000`

Is this behavior intended? is there a way I can completely override the default timeout of `100000` to any value I want?
Also, what is the difference between the timeout passed to `withDriver` in `->withDriver('bolt', $url, null, 3.0)` and the one passed to `withDefaultTransactionConfiguration` in `->withDefaultTransactionConfiguration(TransactionConfiguration::default()->withTimeout(3.0))`? 
It appears that I need to set both of them for it to take effect on transactions, if I set one but not the other it doesn't work for transactions either, and nothing I tried worked for a one-off `run` query as I showed above. is there something I'm doing wrong with the configuration?

_Originally posted by @noureldin-khaled in https://github.com/neo4j-php/neo4j-php-client/issues/64#issuecomment-1009216807_

**Comment by transistive** [comment]
Fixed in version 2.4.0

**Comment by noureldin-khaled** [comment]
Hello again @transistive :wave: 

Unfortunately, I'm running into another issue with the timeout. 
When I set the timeout to a large value for some reason the server responds to me with an empty response before the timeout is reached, which causes the code to throw an exception as it doesn't expect the server to respond with an empty response. So let me show an example

I have the client as follows
```php
$url = 'bolt://test:test@neo4j:7687?database=test'
$this->client = ClientBuilder::create()
                             ->withDriver('bolt', $url, null)
                             ->withDefaultTransactionConfiguration(TransactionConfiguration::default()->withTimeout(150.0))
                             ->withDefaultDriver('bolt')
                             ->build();
```
And then I run this query
```php
$tsx = $this->client->beginTransaction();
$result = $tsx->run("CALL apoc.util.sleep(2000000)");
```
It should timeout after 150 seconds and throw a proper timeout exception, but instead randomly after 60 or 70 or sometimes 100 seconds it gives this error
```
Return value of Bolt\protocol\V3::run() must be of the type array, null returned
```

After tracing the code of the library, it does pass the timeout correctly to the underlying Bolt library, and it gets set properly there. However, for some reason, it doesn't work. In my debugging, I got to this function in the `StreamSocket` class of the Bolt library
```php
    /**
     * Read from connection
     * @param int $length
     * @return string
     * @throws ConnectException
     */
    public function read(int $length = 2048): string
    {
        $res = stream_get_contents($this->stream, $length);

        if (stream_get_meta_data($this->stream)["timed_out"])
            throw ConnectionTimeoutException::createFromTimeout($this->timeout);

        if (empty($res))
            throw new ConnectException('Read error');

        if (Bolt::$debug)
            $this->printHex($res, false);

        return (string)$res;
    }
```
What happens is that `stream_get_contents($this->stream, $length)` gets an empty response (and empty string) before the timeout is reached, and the `stream_get_meta_data($this->stream)` has the `timed_out` equals false since it technically got a response (an empty string), so no Connection exception gets thrown and instead the empty response propagates to the caller in the `V3` class
```php
/**
     * Send RUN message
     * The RUN message requests that a Cypher query is executed with a set of parameters and additional extra data.
     *
     * @link https://7687.org/bolt/bolt-protocol-message-specification-3.html#request-message---run
     * @link https://7687.org/bolt/bolt-protocol-message-specification-4.html#request-message---run
     * @link https://7687.org/bolt/bolt-protocol-message-specification-4.html#request-message---run---44
     * @param string|array ...$args query, parameters, extra
     * @return array
     * @throws Exception
     */
    public function run(...$args): array
    {
        if (empty($args)) {
            throw new PackException('Wrong arguments count');
        }

        $this->write($this->packer->pack(
            0x10,
            $args[0],
            (object)($args[1] ?? []),
            (object)($args[2] ?? [])
        ));
        $message = $this->read($signature);

        if ($signature == self::FAILURE) {
            throw new MessageException($message['message'], $message['code']);
        }

        if ($signature == self::IGNORED) {
            throw new IgnoredException('RUN message IGNORED. Server in FAILED or INTERRUPTED state.');
        }

        return $message;
    }
```
which accordingly returns null hence the error since it should return an array.

Similarily when I run a query that returns something, a similar behavior happens so for example
```php
$tsx = $this->client->beginTransaction();
$result = $tsx->run("CALL apoc.util.sleep(2000000) RETURN 1");
```
It follows a different path in the code but ultimately reaches the same root cause which is that an empty response is returned before the timeout is reached, the error given in that case is
```
Empty bolt result set
``` 

After experimenting with the timeout value a lot, any timeout value up to 70 works perfectly, and an appropriate exception is thrown with no weird behaviors, but any value above that causes this issue, especially large values like 150. 

Worth noting is that the query on neo4j itself wouldn't actually finish when the empty response is returned. So an empty response is returned to the code but the query is still executing on neo4j, and when the exception is thrown and the php process terminates, neo4j terminates the query on its end (as the connection to the client executing the query is terminated).

Also worth noting that this happens when a query is waiting to acquire a lock from another query that is running, it also doesn't timeout properly but instead causes the same error.

I know this might be an issue with the underlying Bolt library, but I thought I would ask you if you have any ideas on how to solve this or why could this be happening? How do I make the timeout work as expected, and why does it respond with an empty string like that? Any help would be much appreciated :pray:

**Comment by transistive** [comment]
Thank you for the massively interesting write-up :monocle_face: 

It's already late where I am, but I will definitely take a look tomorrow! Good stuff. We'll get to the bottom of this!

**Comment by transistive** [comment]
So it was indeed an underlying bolt issue. I have fixed it, but the fix might open pandora's box to a host of other problems. I have therefore requested a review. As soon as it is released, I will up the bolt version requirement as well.

Many thanks once again! Other users also experienced the bug, but I couldn't reproduce it as their input was limited.

If there is an easy way to buy you a coffee, you should send me a link :coffee:

**Comment by noureldin-khaled** [comment]
This is great news! Thank you so much for the quick reply and fix :pray:
Yeah, I have been scratching my head for a while trying to reproduce the scenario consistently. Fingers crossed that the fix will be safe to release :crossed_fingers:

**Comment by noureldin-khaled** [comment]
any updates on this issue? I can see the PR for the fix in the Bolt library got merged, does this mean we can up the bolt version requirement for the client? @transistive

**Comment by transistive** [comment]
Yes, I completely forgot to mention it here! It is fixed in version 2.5.0

Good luck!
